### DISCUSSION TITLE
With Webpack and Gulp, "Cannot use import statement outside a module"

### DISCUSSION THREAD

**Discussion by justinfarrelldev** [babel/babel#14760] [Q&A] [answered]
I checked for similar discussions and my issue seems to be different from #12463 so I figured I'd post it here. 

Exact error (notice the `.tsx` extension on the script as well): 
```
Uncaught SyntaxError: Cannot use import statement outside a module (at e0e4a24c64b9937fd5aa.tsx:1:1)
```

Tooling: 
```
React
TypeScript
Gulp (with a gulpfile.ts)
Webpack (with a webpack.config.ts)
```

When I try to run the dev server, the script I have attached doesn't seem to be imported as a module.

I tried putting `"type": "module"` into the tsconfig.json, however if I do that then Gulp stops working (`"unknown file extension: .ts"`). If I feed it an environment variable to force Node to use commonjs then reload, then the same "Cannot use import...." error is still visible in the browser. 

The other thing I tried is adding `type: "module"` to the script tag in the HTML, however that then leads to the mime type issue below:  

```
Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "application/octet-stream". Strict MIME type checking is enforced for module scripts per HTML spec.
```

I simply don't know what to do or what could be wrong. Any help would be much appreciated! 

package.json: 
```
{
  "version": "1.0.0",
  "main": "index.tsx",
  "scripts": {
    "test": "jest --maxWorkers=50%",
    "start": "gulp -LL start"
  },
  "dependencies": {
    "@emotion/react": "^11.9.3",
    "@emotion/styled": "^11.9.3",
    "@material-ui/core": "^4.5.1",
    "@material-ui/icons": "^4.5.1",
    "@material-ui/lab": "^4.0.0-alpha.39",
    "@material-ui/pickers": "^3.2.7",
    "@mui/icons-material": "^5.6.2",
    "@mui/material": "^5.9.0",
    "css-loader": "^6.7.1",
    "file-loader": "^6.2.0",
    "html-loader": "^4.1.0",
    "html-webpack-plugin": "^5.5.0",
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "shelljs": "^0.8.5",
    "style-loader": "^3.3.1",
    "ts-loader": "^9.3.1",
    "typescript": "^4.7.4"
  },
  "devDependencies": {
    "@babel/core": "^7.18.6",
    "@babel/plugin-transform-modules-commonjs": "^7.18.6",
    "@babel/polyfill": "^7.12.1",
    "@babel/preset-env": "^7.18.6",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@types/gulp": "^4.0.9",
    "@types/material-ui": "^0.21.12",
    "babel-loader": "^8.2.5",
    "babel-preset-react": "^6.24.1",
    "gulp": "^4.0.2",
    "gulp-param": "^1.0.3",
    "gulp-typescript": "^6.0.0-alpha.1",
    "ts-node": "^10.8.2",
    "webpack": "^5.73.0",
    "webpack-cli": "^4.10.0",
    "webpack-dev-server": "^4.9.3",
    "yargs": "^17.5.1"
  }
}
```

webpack.config.ts (with the Babel config): 
```
import path from "path";
import HtmlWebpackPlugin from "html-webpack-plugin";

module.exports = {
  output: { path: path.join(__dirname, "build"), filename: "index.bundle.js" },
  mode: process.env.NODE_ENV || "development",
  resolve: {
    extensions: [".tsx", ".ts", ".jsx", ".js", ".html", ".less"],
  },
  module: {
    rules: [
      {
        test: /\.(ts|tsx|js|jsx)$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: [
              "@babel/preset-typescript",
              "@babel/preset-react",
              "@babel/preset-env",
            ],
            plugins: ["@babel/plugin-proposal-class-properties"],
            sourceType: "module",
          },
        },
      },
      {
        test: /\.(css|scss)$/,
        exclude: /node_modules/,
        use: ["style-loader", "css-loader"],
      },
      {
        test: /\.(jpg|jpeg|png|gif|mp3|svg)$/,
        exclude: /node_modules/,
        use: ["file-loader"],
      },
      {
        test: /\.(html|htm)$/,
        exclude: /node_modules/,
        use: ["html-loader"],
      },
    ],
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: path.join(__dirname, process.argv[4]),
    }),
  ],
};
```

tsconfig.json: 
```
{
    "compilerOptions": {
        "jsx": "react-jsx",
        "esModuleInterop": true
    },
    "type": "module"
}
```

I have been triggering the webpack serve command inside of my gulpfile using shelljs as follows: 
```
  shelljs.exec(`webpack serve --entry ./src/${moduleToStart}/index.html`);
```

Thank you again!

**Comment by JLHwung** [comment, marked as answer]
> Uncaught SyntaxError: Cannot use import statement outside a module (at e0e4a24c64b9937fd5aa.tsx:1:1)

Can you post this file? Does the import statement come from your app? Or from your node modules?

**Reply by justinfarrelldev** [reply]
> Can you post this file? Does the import statement come from your app? Or from your node modules?

```
import React from "react";
import ReactDOM from "react-dom";
import { Login } from "./main";
ReactDOM.render( /*#__PURE__*/React.createElement(Login, null), document.getElementById("root"));
```

The import that errors is the first line just importing React from the "react" module. Thank you for the quick reply

EDIT: 
Original file (index.tsx): 
```
import React from "react";
import ReactDOM from "react-dom";
import { Login } from "./main";

ReactDOM.render(<Login />, document.getElementById("root"));
```

**Reply by JLHwung** [reply]
> Uncaught SyntaxError: Cannot use import statement outside a module (at e0e4a24c64b9937fd5aa.tsx:1:1)

Where is the error from? Is it thrown when you open your browser? Or thrown in the terminal when you are bundling your app?

**Reply by justinfarrelldev** [reply]
> Where is the error from? Is it thrown when you open your browser? Or thrown in the terminal when you are bundling your app?

It's thrown in the console tab of the Chrome dev tools: 
<img width="897" alt="image" src="https://user-images.githubusercontent.com/42923203/179078190-949e9a06-7988-407b-b8ee-eeb8f1d1b9fa.png">

The same error in Sources: 
<img width="419" alt="image" src="https://user-images.githubusercontent.com/42923203/179078316-a89ea129-46fb-48d3-8474-774bd6b93e48.png">

**Reply by JLHwung** [reply]
It seems to me the webpack is not bundling your sources at all. Generally it is a configuration issue but the config you have posted seems correct to me. Can you share a reproduction repo?

**Reply by justinfarrelldev** [reply]
> It seems to me the webpack is not bundling your sources at all. Generally it is a configuration issue but the config you have posted seems correct to me. Can you share a reproduction repo?

Created a reproduction repo at: https://github.com/justinfarrelldev/reproduction

**Reply by JLHwung** [reply]
Thanks for the repo, it helps a lot.

1) in the HtmlWebpackPlugin template, one is not supposed to add a script tag: The HtmlWebpackPlugin will do it for you, so https://github.com/justinfarrelldev/reproduction/blob/eb11cae9a2fd7e0b424286e3d79931b0d2e873ec/src/login/index.html#L20 should be removed

2) The `entry` of webpack should be resolvable to any resources that would be transformed into a `.js`. So the invoking command should be `webpack serve --entry ./src/${moduleToStart}` and webpack will resolve `index.tsx` for you. But we don't have any loader that transforms html to js.

3) Since the `entry` is modified, we should add `index.html` to the `template` of the `HtmlWebpackPlugin`.

Here is the whole patch to make it work:

```diff
diff --git a/gulpfile.ts b/gulpfile.ts
index 5d5f983..0566355 100644
--- a/gulpfile.ts
+++ b/gulpfile.ts
@@ -24,7 +24,7 @@ gulp.task('start', async () => {
     console.error(`Module "${moduleToStart}" not found in module map.`);
     return;
   }
-  shelljs.exec(`webpack serve --entry ./src/${moduleToStart}/index.html`);
+  shelljs.exec(`webpack serve --entry ./src/${moduleToStart}`);
 });
 
 exports.default = defaultTask;
diff --git a/src/login/index.html b/src/login/index.html
index ef810aa..2da6169 100644
--- a/src/login/index.html
+++ b/src/login/index.html
@@ -17,6 +17,5 @@
       >HTML shows up, but check the console. This is just plain HTML - React
       components don't render</h1
     >
-    <script type="text/javascript" src="index.tsx"></script>
   </body>
 </html>
diff --git a/webpack.config.ts b/webpack.config.ts
index 6154b30..e553ff2 100644
--- a/webpack.config.ts
+++ b/webpack.config.ts
@@ -43,7 +43,7 @@ module.exports = {
   },
   plugins: [
     new HtmlWebpackPlugin({
-      template: path.join(__dirname, process.argv[4]),
+      template: path.join(__dirname, process.argv[4], "index.html"),
     }),
   ],
 };
```

**Reply by justinfarrelldev** [reply]
> Here is the whole patch to make it work:

Thank you so much, it worked! Marking it as answered and linking back to this from my reproduction 😃 

Also thank you for the thorough answer, I will keep that in mind going forward. I'm ecstatic that this ended up working!
